src/transform.js

```diff
@@ -250,12 +250,18 @@ const transform = {
   NullableTypeAnnotation: {
     exit(path) {
       const { typeAnnotation } = path.node;
+
+      // conditionally unwrap TSTypeAnnotation nodes
+      const unwrappedType = t.isTSTypeAnnotation(typeAnnotation)
+        ? typeAnnotation.typeAnnotation
+        : typeAnnotation;
+
       path.replaceWith(
         t.tsUnionType([
-          // conditionally unwrap TSTypeAnnotation nodes
-          t.isTSTypeAnnotation(typeAnnotation)
-            ? typeAnnotation.typeAnnotation
-            : typeAnnotation,
+          // conditionally wrap function types in parens
+          t.isTSFunctionType(unwrappedType)
+            ? t.tsParenthesizedType(unwrappedType)
+            : unwrappedType,
```

Comment:
I'm not really sure why the babel AST has a specific node to parenthesize types, but that's apparently how this should be structured.  I checked astexplorer.net.